FirebaseMLModelDownloader/Sources/FileDownloader.swift

```diff
@@ -0,0 +1,114 @@
+  )
+  /// Successful download completion handler.
+  private var completion: FileDownloader.CompletionHandler?
+  /// Download progress handler.
+  private var progressHandler: FileDownloader.ProgressHandler?
+  /// Failed download completion handler.
+  private var downloadErrorHandler: FileDownloader.DownloadErrorHandler?
+  /// Configuration error handler.
+  private var configurationErrorHandler: FileDownloader.ConfigurationErrorHandler?
+
+  init(conditions: ModelDownloadConditions) {
+    self.conditions = conditions
+    configuration = URLSessionConfiguration.ephemeral
+    /// Wait for network connectivity, if unavailable.
+    if #available(iOS 11.0, macOS 10.13, macCatalyst 13.0, tvOS 11.0, watchOS 4.0, *) {
+      self.configuration.waitsForConnectivity = true
+      /// Wait for 10 minutes.
+      self.configuration.timeoutIntervalForResource = 600
+    }
+    configuration.allowsCellularAccess = conditions.allowsCellularAccess
+  }
+
+  func downloadFile(with url: URL,
+                    progressHandler: @escaping (Int64, Int64) -> Void,
+                    configurationErrorHandler: @escaping (Error) -> Void,
+                    downloadErrorHandler: @escaping (Error) -> Void,
+                    completion: @escaping (URLResponse, URL) -> Void) {
+    self.completion = completion
+    self.progressHandler = progressHandler
+    self.configurationErrorHandler = configurationErrorHandler
+    self.downloadErrorHandler = downloadErrorHandler
+    let downloadTask = downloadSession.downloadTask(with: url)
+    downloadTask.resume()
+    self.downloadTask = downloadTask
+  }
+}
+
+/// Extension to handle delegate methods.
+extension ModelFileDownloader: URLSessionDownloadDelegate {
+  /// Handle client-side errors.
+  func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+    guard let error = error else { return }
+    /// Unable to resolve hostname or connect to host.
+    downloadErrorHandler?(error)
+  }
+
+  func urlSession(_ session: URLSession,
+                  downloadTask: URLSessionDownloadTask,
+                  didFinishDownloadingTo location: URL) {
+    guard let response = downloadTask.response else { return }
+    completion?(response, location)
+  }
+
+  func urlSession(_ session: URLSession,
+                  downloadTask: URLSessionDownloadTask,
+                  didWriteData bytesWritten: Int64,
+                  totalBytesWritten: Int64,
+                  totalBytesExpectedToWrite: Int64) {
+    progressHandler?(totalBytesWritten, totalBytesExpectedToWrite)
+  }
+
+  func urlSession(_ session: URLSession, taskIsWaitingForConnectivity task: URLSessionTask) {
+    // TODO: Handle waiting for connectivity, if needed.
+  }
+
+  func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
+    guard let error = error else { return }
+    /// Unable to resolve hostname or connect to host.
+    configurationErrorHandler?(error)
```

Comment:
nit: I think this error can be reported to `downloadErrorHandler`

Reply:
Ah but what if we want to handle download errors and configuration errors differently in the caller?

Reply:
I would suggest to define an error type to differentiate different types of errors reported by the `FileDownloader`, e.g.

```
enum FileDownloaderError: Error { 
  case sessionInvalidated(Error)
  case unexpectedResponseType
  case networkError(Error)
  ...
}
```

In this case the API will be a bit cleaner but the client has a chance differentiate types of errors if needed.

nit: Additionally, the completion success and error handlers may be merged into one, but it is up to you, e.g.
```
struct FileDownloaderResponse {
    var urlResponse: HTTPURLResponse
    var fileURL: URL
  }

protocol FileDownloader {
  func downloadFile(with url: URL, completion: (Result<FileDownloaderResponse, Error>)->Void)
}

```

Reply:
Great idea! I made this change.